presto-main/src/main/java/com/facebook/presto/cost/VariableStatsEstimate.java

```diff
@@ -237,9 +265,16 @@ public Builder setDistinctValuesCount(double distinctValuesCount)
             return this;
         }
 
+        public Builder setHistogram(ConnectorHistogram histogram)
+        {
+            this.histogram = Optional.of(histogram);
+            return this;
+        }
+
         public VariableStatsEstimate build()
         {
-            return new VariableStatsEstimate(lowValue, highValue, nullsFraction, averageRowSize, distinctValuesCount);
+            return new VariableStatsEstimate(lowValue, highValue, nullsFraction, averageRowSize, distinctValuesCount,
+                    histogram.orElseGet(() -> new UniformDistributionHistogram(lowValue, highValue)));
```

Comment:
Here we will set and use the default histogram even without the session property. Is that intended or should we make sure that we use a noop version without?

Reply:
This is a good point. It will populate the histogram in `VariableStatsEstimate` without the property set. However, all of the code which actually uses the histograms for calculations is gated by the property here:

https://github.com/prestodb/presto/pull/21236/commits/838628b0574c3606302ba9e9f219b53c201d3c70#diff-3e8f0b799abcb81b4285a5d98fde631da64d29d6fd95e952b3583ffa949301f6R170

I think the tough thing about changing this is that we would need to pass `Session` everywhere in order to check the property to determine whether or not the histogram should be included on the stats objects. This isn't the only location in the code where it includes histograms. There are others such as [here](https://github.com/prestodb/presto/pull/21236/commits/838628b0574c3606302ba9e9f219b53c201d3c70#diff-22fa802402957ec23b9dfd10c46f8058045665fa29a8e9abe67398ae6be6f071R136) and [here](https://github.com/prestodb/presto/pull/21236/commits/838628b0574c3606302ba9e9f219b53c201d3c70#diff-22fa802402957ec23b9dfd10c46f8058045665fa29a8e9abe67398ae6be6f071R287-R290). As long as the calculation done in `ComparisonStatsCalculator` is the only location that histograms are used, I think it is safe to just add the histogram here by default

Reply:
What if we store session in the implementations of the **ConnectorHistorgram** and throw exceptions when used without?

Reply:
Even if you stored the session in the histogram, it would still require changing all of the creation sites of the histogram (and hence the creation sites of `VariableStatisticsEstimate`). Also, the property controlling histogram use is defined in `SystemSessionProperties`. I don't think there is much benefit doing it that way.

I went through a few options and I think the one that will touch the smallest amount of code will require at least having conditionals at all call sites where the histograms are created/added to stats estimates in the optimizer. From a cursory search it is about ~10 locations, plus a good handful of function signature changes to pass the session. Other options such as changing `VariableStatisticEstimate` to store a `Session` member will require 1-200 of LoC change.

Aside from a future change accidentally using the histogram without gating it by the flag, is there any other specific concern with including the histograms on the `VariableStatsEstimate` even when the feature is disabled?

Reply:
Okay I see, nope that was the only concern.